### PR TITLE
Kc change

### PR DESCRIPTION
**Pull request by nolantellis**

### CODE REVIEW

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/CustomOIDCFilter.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    if (requestPath.startsWith("/login.zul")) {
+      String redirecturi = getRedirectUri(request, requestPath);
+      response.sendRedirect(redirecturi);
+      return;
+    }
+
+    boolean isExistingSession = existingSession == null ? false
+        : existingSession.getAttribute("USER") == null ? false : true;
+
+    System.out.println("Auth-path=" + requestPath);
```
Comment (nolantellis) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/CustomOIDCFilter.java]: remove this

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/security/KeycloakLoginUrlAuthenticationEntryPoint.java [hunk 2]
```diff
@@ -21,158 +21,146 @@
...
+      LOGGER.trace("[ Keycloak SSO turned off ]");
+
+      String requestUriStr = httpServletRequest.getRequestURL().toString().trim();
+      HttpSession session = httpServletRequest.getSession(false);
+      URI uri = null;
+      try {
+        uri = new URI(requestUriStr);
+      } catch (URISyntaxException e) {
+        // TODO Auto-generated catch block
+        e.printStackTrace();
```
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/security/KeycloakLoginUrlAuthenticationEntryPoint.java]: Would be good to log exception as error.

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/ApromoreFilterConfig.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+
+
+  public ApromoreFilterConfig(ServletContext context) {
+    this.context = context;
+  }
+
+  @Override
+  public String getFilterName() {
+    // TODO Auto-generated method stub
```
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/ApromoreFilterConfig.java]: [Minor] remove auto-generated comments

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/AuthDetailsFilter.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+  @Autowired
+  ManagerService managerService;
+
+  @Value("${site.useKeycloakSso}")
+  private boolean useKeyCloak = true;
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+
```
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/AuthDetailsFilter.java]: [Minor] Remove blank lines.

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/CustomOIDCFilter.java (hunk 1 above)
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/CustomOIDCFilter.java]: Use logger rather than System.out

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/ApromoreFilterConfig.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    return context;
+  }
+
+  @Override
+  public String getInitParameter(String name) {
+    switch (name) {
+      case "keycloak.config.skipPattern":
+        return null;
+
```
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/ApromoreFilterConfig.java]: [Minor] remove blank lines

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/CustomOIDCFilter.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+
+    skipPattern = FilterRegexUtil.getPattern();
+
+    File f = new File(System.getProperty("karaf.etc"), "/config/" + "keycloak.json");
+    try {
+      FileInputStream keyCloakJsonFile = new FileInputStream(f);
```
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/CustomOIDCFilter.java]: Use try-with-resources for IO resources

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/AuthDetailsFilter.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    HttpServletRequest httpRequest = (HttpServletRequest) request;
+    RefreshableKeycloakSecurityContext context = (RefreshableKeycloakSecurityContext) request
+        .getAttribute("org.keycloak.KeycloakSecurityContext");
+
+    AccessToken token = context.getToken();
+
+    String userName = token.getPreferredUsername();
+    userName = userName == null ? token.getEmail() : userName;
+    String email = token.getEmail();
+    String givenName = token.getGivenName();
```
Comment (fsmicdev) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/AuthDetailsFilter.java]: These two statements (givenName and lastName retrieval) can be moved after line 96 as only utilised there, or used in-line on line 97 directly.  Reduces scope, and provides a slight performance improvement in the alternate use case (when userType != null).

**Review by fsmicdev — CHANGES_REQUESTED** [review]
Minor changes, as indicated.

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/ApromoreFilterConfig.java (hunk 3 above)
Comment (nolantellis) [reply Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/ApromoreFilterConfig.java]: Will delete this file. Not used

**Review by fsmicdev — APPROVED** [review]

File: site.properties [hunk 8]
```diff
@@ -203,7 +203,7 @@ jpa.generateDDL      = false
 
 ## Liquibase User Mysql
 liquibase.jdbc.username = liquibase_user
-liquibase.jdbc.password = 7fHJV41fpJ
+liquibase.jdbc.password = MAcri
```
Comment (raboczi) [inline site.properties]: Is this password change intentional?  It prevents existing developers from starting Apromore unless they go into MySQL to change the password there.  If you do think changing the password is worth doing, it also has to be updated on line 59 of README.md.

FWIW, I like the idea of jdbc.password and liquibase.jdbc.password having different values, just to remind people that they don't have to (shouldn't) be the same.

It seems like reverting site.properties entirely is probably viable.

File: site.properties [hunk 9]
```diff
@@ -148,7 +148,7 @@ portal.menuitemorder.File = Upload,Download,Create data pipeline,Export log as C
 contact.email = support@apromore.atlassian.net
 
 # Calendar
-calendar.enable = false
+calendar.enable = true
```
Comment (raboczi) [inline site.properties]: This has nothing to do with Keycloak and shouldn't be part of this PR.

File: core-assemblies/core-features/src/main/feature/feature.xml [hunk 10]
```diff
@@ -136,6 +136,16 @@
...
     <bundle>mvn:com.nimbusds/nimbus-jose-jwt/9.7</bundle>
+    <bundle>mvn:org.keycloak/keycloak-servlet-filter-adapter/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-core/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-common/14.0.0</bundle>    
+    <bundle>mvn:org.keycloak/keycloak-adapter-spi/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-servlet-adapter-spi/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-adapter-core/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-servlet-filter-adapter/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-authz-client/14.0.0</bundle>
+    <bundle>mvn:org.keycloak/keycloak-osgi-adapter/14.0.0</bundle>
```
Comment (raboczi) [inline core-assemblies/core-features/src/main/feature/feature.xml]: Update the keycloak.version property in ApromoreCore/pom.xml and use ${keycloak.version} instead of 14.0.0

File: core-assemblies/apromore-core/pom.xml [hunk 11]
```diff
@@ -244,6 +244,7 @@
               <destinationFile>${project.build.directory}/assembly/etc/site.cfg</destinationFile>
             </configuration>
           </execution>
+
```
Comment (raboczi) [inline core-assemblies/apromore-core/pom.xml]: This file has no meaningful edits and ought to be reverted.

File: Apromore-Core-Components/Apromore-Portal/src/main/webapp/WEB-INF/zk.xml [hunk 12]
```diff
@@ -400,15 +400,13 @@
...
 
-    <listener>
+   <!-- <listener>
             <listener-class>org.apromore.portal.ApromoreZKListener</listener-class>
-    </listener>
+    </listener> -->
 
-    <listener>
-            <listener-class>org.apromore.portal.LoggingZKListener</listener-class>
-    </listener>
```
Comment (raboczi) [inline Apromore-Core-Components/Apromore-Portal/src/main/webapp/WEB-INF/zk.xml]: ApromoreZKListener had two functions: it listened for ZK Executions (servlet requests) to do the JWT checking (which is valid for you to remove here).  But it also listened for ZK WebApps (servlet contexts) to override the cookie configuration.  Removing the entire thing probably breaks the fix for correcting the cookie flags.  We could tweak Nginx to cover for this, but it's better if the app is correct on its own without Nginx.

The LoggingZKListener should not be removed.

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/FilterRegexUtil.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+package org.apromore.portal.servlet.filter;
+
+import java.io.File;
+import java.util.regex.Pattern;
+
+public class FilterRegexUtil {
+
+  static Pattern p = Pattern.compile("^/(zkau|zkau/.*|.*\\.css|.*\\.js"
+      + "|.*\\.(svg|jpg|jpeg|png)|themes/.*|favicon\\.ico" + "|css/.*|" + ".*/js/.* |.*/fonts/.*)",
+      Pattern.DOTALL);
```
Comment (raboczi) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/FilterRegexUtil.java:31]: This seems like it belongs in a configuration file rather than being hardcoded.

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/security/KeycloakLoginUrlAuthenticationEntryPoint.java [hunk 14]
```diff
@@ -21,158 +21,146 @@
...
+    } else {
+      LOGGER.trace("[ Keycloak SSO turned off ]");
+
+      String requestUriStr = httpServletRequest.getRequestURL().toString().trim();
+      HttpSession session = httpServletRequest.getSession(false);
+      URI uri = null;
+      try {
+        uri = new URI(requestUriStr);
+      } catch (URISyntaxException e) {
+        LOGGER.error("Exception " + ExceptionUtils.getStackTrace(e));
```
Comment (raboczi) [inline Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/security/KeycloakLoginUrlAuthenticationEntryPoint.java]: `LOGGER.error("Bad request URI", e);`

File: Apromore-Core-Components/Apromore-Portal/pom.xml [hunk 15]
```diff
@@ -378,6 +385,24 @@
             <artifactId>guava</artifactId>
         </dependency>
 
+		<dependency>
+		    <groupId>org.keycloak</groupId>
+		    <artifactId>keycloak-servlet-filter-adapter</artifactId>
+		    <version>14.0.0</version>
```
Comment (raboczi) [inline Apromore-Core-Components/Apromore-Portal/pom.xml]: Update keycloak.version in ApromoreCore/pom.xml and use ${keycloak.version}

File: Apromore-Core-Components/Apromore-Portal/pom.xml [hunk 16]
```diff
@@ -378,6 +385,24 @@
...
+	          <groupId>commons-logging</groupId>
+	          <artifactId>commons-logging</artifactId>
+	        </exclusion>
+	      </exclusions> 
+		</dependency>
+		
+		<dependency>
+    		<groupId>org.keycloak</groupId>
+    		<artifactId>keycloak-osgi-adapter</artifactId>
+    		<version>14.0.0</version>
```
Comment (raboczi) [inline Apromore-Core-Components/Apromore-Portal/pom.xml]: Update keycloak.version in ApromoreCore/pom.xml and use ${keycloak.version}

**Review by raboczi — CHANGES_REQUESTED** [review]

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/security/KeycloakLoginUrlAuthenticationEntryPoint.java (hunk 14 above)
Comment (raboczi) [reply Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/security/KeycloakLoginUrlAuthenticationEntryPoint.java]: Done.

File: Apromore-Core-Components/Apromore-Portal/src/main/webapp/WEB-INF/zk.xml (hunk 12 above)
Comment (raboczi) [reply Apromore-Core-Components/Apromore-Portal/src/main/webapp/WEB-INF/zk.xml]: Done.

File: core-assemblies/apromore-core/pom.xml (hunk 11 above)
Comment (raboczi) [reply core-assemblies/apromore-core/pom.xml]: Done.

File: site.properties (hunk 9 above)
Comment (raboczi) [reply site.properties]: Reverted.

File: site.properties (hunk 8 above)
Comment (raboczi) [reply site.properties]: Reverted.

File: Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/FilterRegexUtil.java (hunk 13 above)
Comment (nolantellis) [reply Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/FilterRegexUtil.java:31]: will be done in springboot

Comment (raboczi) [reply Apromore-Core-Components/Apromore-Portal/src/main/java/org/apromore/portal/servlet/filter/FilterRegexUtil.java:31]: Fair enough.

**Review by raboczi — APPROVED** [review]
